amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/Scheduler.java

```diff
@@ -458,6 +466,90 @@ void runProcessLoop() {
+            }
+        }
+        return allParentShardLeasesDeleted;
+    }
+
+    private void cleanupCompletedShards(Collection<ShardInfo> completedShards)
+            throws DependencyException, ProvisionedThroughputException, InvalidStateException {
+        
+        for (ShardInfo completedShard : completedShards) {
+            final Lease lease = leaseCoordinator.getCurrentlyHeldLease(ShardInfo.leaseKeyProvider.apply(completedShard));
+            final Set<String> childShardKeys = lease.childShardIds().stream().map(c -> ShardInfo.getLeaseKey(completedShard, c)).collect(Collectors.toSet());
+
+            if (CollectionUtils.isNullOrEmpty(lease.childShardIds())) {
+                throw new IllegalStateException("No child shard state found in lease " + lease + " - not deleting lease.");
+            } else {
+                final Set<String> processedChildShardLeases = new HashSet<>();
+
+                for (String childShard : lease.childShardIds()) {
+                    final String childShardLeaseKey = ShardInfo.getLeaseKey(completedShard, childShard);
+                    final Lease childShardLease = Optional.ofNullable(leaseCoordinator.leaseRefresher().getLease(childShardLeaseKey)).orElseThrow(
+                            () -> new IllegalStateException("Current lease " + childShardLeaseKey + " not found in lease table."));
+
+                    if (!childShardLease.checkpoint().equals(ExtendedSequenceNumber.TRIM_HORIZON)
+                            && !childShardLease.checkpoint().equals(ExtendedSequenceNumber.AT_TIMESTAMP)) {
+                        processedChildShardLeases.add(childShardLease.leaseKey());
+                    }
+                }
+
+                if (allParentShardLeasesDeleted(lease, completedShard) && Objects.equals(childShardKeys, processedChildShardLeases)) {
+                    log.info("Deleting lease {} as it has been completely processed and processing of child shards has begun.",
+                            lease);
+                    leaseCoordinator.leaseRefresher().deleteLease(lease);
+                    leaseCoordinator.dropLease(lease);
+                }
+            }
+        }
+    }
+
+    private void cleanupGarbageShards(Collection<ShardInfo> garbageShards) throws DependencyException, ProvisionedThroughputException, InvalidStateException {
+        for (ShardInfo garbageShard : garbageShards) {
+            final Lease lease = leaseCoordinator.getCurrentlyHeldLease(ShardInfo.leaseKeyProvider.apply(garbageShard));
+            if (allParentShardLeasesDeleted(lease, garbageShard)) {
+                log.info("Deleting lease {} as it is not present in the stream.");
+                leaseCoordinator.leaseRefresher().deleteLease(lease);
+                leaseCoordinator.dropLease(lease);
+            }
+        }
+    }
+
+    private List<ShardInfo> getGarbageShards(List<ShardInfo> shardInfoAssignments) throws ExecutionException, InterruptedException {
```

Comment:
shardInfoAssignments does it contains the shards which are owned by KCL worker ? 

Also, making TRIM_HORIZON call here repeatedly might increase load on the service side. Can we optimize to make for shards which are not making progress ? 

This might increase our trim reads call as this goes to the BE for all customers with 24H of retention

Reply:
> shardInfoAssignments does it contains the shards which are owned by KCL worker ?
Yes, this is an in-memory mapping of the leases each worker owns, and they will be periodically trying to renew this.

> Also, making TRIM_HORIZON call here repeatedly might increase load on the service side. Can we optimize to make for shards which are not making progress ?

Will this be less intensive if we request LATEST? Will look into doing this for just shards which aren't making progress.

Reply:
We are now just using S2S/GetRecords response in latest revision of PR.